picoquictest/skip_frame_test.c

```diff
@@ -799,33 +804,105 @@ int logger_test()
+
     if (f == NULL) {
         DBG_PRINTF("failed to open file:%s\n", binlog_test_file);
         ret = -1;
-    } else {
-        for (size_t i = 0; i < nb_test_skip_list; i++) {
-            picoquic_binlog_frames(f, test_skip_list[i].val, test_skip_list[i].len);
-        }
-        (void)picoquic_file_close(f);
+    } else if (quic == NULL) {
+        DBG_PRINTF("%s", "Cannot create QUIC context\n");
+        ret = -1;
+    } else if (ret_bin != 0 || ret_qlog != 0) {
+        DBG_PRINTF("%s", "Cannot set the log ref file name.\n");
+        ret = -1;
     }
+    else {
+        quic->f_binlog = f;
+        
+        picoquic_set_default_spinbit_policy(quic, picoquic_spinbit_null);
 
-    if (ret == 0) {
+        struct sockaddr_in saddr;
+        memset(&saddr, 0, sizeof(struct sockaddr_in));
+        picoquic_cnx_t* cnx = picoquic_create_cnx(quic, srce_cid, dest_cid, (struct sockaddr*) & saddr,
+            simulated_time, 0, "test-sni", "test-alpn", 1);
 
-        char log_test_ref[512];
-        ret = picoquic_get_input_path(log_test_ref, sizeof(log_test_ref), picoquic_test_solution_dir, BINLOG_TEST_REF);
+        if (cnx == NULL) {
+            DBG_PRINTF("%s", "Cannot create QUIC CNX context\n");
+            ret = -1;
+        } else {
+            for (size_t i = 0; i < nb_test_skip_list; i++) {
+
+                picoquic_packet_header ph;
+                memset(&ph, 0, sizeof(ph));
+
+                ph.ptype = picoquic_packet_1rtt_protected;
+                ph.pn64 = i;
+                ph.dest_cnx_id = srce_cid;
+                ph.srce_cnx_id = dest_cid;
+
+                ph.offset = 0;
+                ph.payload_length = test_skip_list[i].len;
+
+                binlog_packet(quic->f_binlog, &srce_cid, 0, 0, &ph, test_skip_list[i].val, test_skip_list[i].len);
+            }
+
+            picoquic_delete_cnx(cnx);
+        }
+    }
+
+    picoquic_free(quic);
 
+    if (ret == 0) {
+        ret = picoquic_test_compare_binary_files(binlog_test_file, log_test_ref);
         if (ret != 0) {
-            DBG_PRINTF("%s", "Cannot set the log ref file name.\n");
-        } else {
-            ret = picoquic_test_compare_binary_files(binlog_test_file, log_test_ref);
+            DBG_PRINTF("%s", "Unexpected content in binary log file.\n");
+        }
+        else {
+            /* Convert to QLOG and verify */
+            uint64_t log_time = 0;
+            FILE* f_binlog = picoquic_open_cc_log_file_for_read(binlog_test_file, &log_time);
+            ret = qlog_convert(&srce_cid, f_binlog, binlog_test_file, ".");
+            if (ret != 0) {
+                DBG_PRINTF("%s", "Cannot convert the binary log into QLOG.\n");
+            }
+            else {
+                /* When changing the reference QLOG file please verify the new file at:
+                   https://qvis.edm.uhasselt.be/#/files */
+                ret = picoquic_test_compare_text_files(qlog_test_file, qlog_test_ref);
+            }
         }
```

Comment:
That code will only perform the Qlog comparison if the bin file binaries are identical. That's a good way to test qlog.c, but that does not solve the "binary debug" problem. Suppose that we change the log content for any reason: maybe we have added a new frame type, maybe we have changed the format of some frame. The binaries will be different. If they are different "in the way that we expect", then life is good, we should just replace the reference bin log by the new version. But if they are different in a bad way, we should fix the code and leave the reference untouched.

Maybe what we need is a structured comparison. Read the binary log and the reference record by record, and report on the differences. But failing that, we could run the qlog comparison all the time, even if the binaries are different. Then when the test fail we can inspect the qlog file to understand what changed.

Reply:
I am ok with removing the binary comparison all together. Binary to QLOG conversion should be bijective one day anyhow.

Reply:
Works for me.